R/plugin.R

```diff
@@ -0,0 +1,214 @@
+  } else {
+    path <- getwd()
+    root <- detect_orderly_interactive_path(path)
+    orderly_config <- orderly_root(root$path, FALSE)$config
+    env <- orderly_plugin_environment(name)
+    src <- path
+    parameters <- current_orderly_parameters(src, env)
+  }
+  check_plugin_enabled(name, orderly_config)
+  config <- orderly_config$plugins[[name]]$config
+  list(is_active = is_active, path = path, config = config, env = env,
+       src = src, parameters = parameters)
+}
+
+
+##' Add plugin-specific metadata to a running packet. This will take
+##' some describing. You accumulate any number of bits of metadata
+##' into arbitrary fields, and then later on serialise these to json.
+##'
+##' @title Add metadata from plugin
+##'
+##' @param name The name of the plugin; must be the same as used in
+##'   [orderly3::orderly_plugin_register] and
+##'   [orderly3::orderly_plugin_context]
+##'
+##' @param field The name of a field to add the data to. This is
+##'   required even if your plugin only produces one sort of data, in
+##'   which case you can remove it later on within your serialisation
+##'   function.
+##'
+##' @param data Arbitrary data to be added to the currently running
+##'   packet
+##'
+##' @return Nothing, called only for its side effects
+##'
+##' @export
+orderly_plugin_add_metadata <- function(name, field, data) {
+  assert_scalar_character(name)
+  assert_scalar_character(field)
+  p <- get_active_packet()
+  if (!is.null(p)) {
+    check_plugin_enabled(name, p$orderly3$config)
+    p$orderly3$plugins[[name]][[field]] <-
+      c(p$orderly3$plugins[[name]][[field]], list(data))
+  }
+}
+
+
+## This is a real trick, and is only used in the case where the report
+## is being run interactively, and in that case the correct thing is
+## *almost certainly* the global environment, as the user has to do
+## some tricks to stop that being the case; for example running
+##
+##   source("orderly.R", local = TRUE)
+##
+## We want to find the environment that corresponds to the top level
+## environment for orderly; that will be the one that called the
+## plugin function. So we'll have a stack of frames (corresponding to
+## the environments on the call stack) with *parents* that look like
+## this, outermost first:
+##
+## - (anything else)
+## - (calling environment) <-- this is what we're looking for
+## - (plugin)
+## - (orderly3; orderly_plugin_context)
+## - (orderly3; from orderly_plugin_environment)
+##
+## so we loop down the stack looking for the first call to a function
+## in the plugin package, then take the frame *above* that.
+orderly_plugin_environment <- function(name) {
+  frames <- sys.frames()
+  for (i in seq_along(frames)[-1]) {
+    if (environmentName(parent.env(frames[[i]])) == name) {
+      return(frames[[i - 1]])
+    }
+  }
+  ## This error should never surface if the plugin is configured correctly
+  stop("Could not determine plugin calling environment safely - please report")
+}
```

Comment:
This will be useful actually when trying to get subqueries working with `orderly_depends`. We want to be able to do something like

```
orderly3::orderly_depends("depends", "latest",
                          c(depends_graph.png = "graph.png"),
                          as = "depends")
orderly3::orderly_depends("explicit", "usedby({depends})",
                          c(explicit_graph.png = "mygraph.png"))
```

which is fine if running via `orderly_run` as we can use the active packet environment but if running interactively we're going to need to determine the environment like this. Any thoughts on if we should try and make this generic or just add a similar function?

Reply:
Oh actually maybe we just need the `parent.frame` in the context I am talking about!

Reply:
I was discussing this with Alex - I think that the whole context thing might want exposing more internally. It would be useful for the global dependencies too